cmd/snap-confine/mount-support.c

```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
 	return false;
 }
 
+static struct sc_mount *sc_homedir_mounts(const struct sc_invocation *inv)
+{
+	if (inv->homedirs == NULL) {
+		// Return empty array, but never NULL as functions rely
+		// on the mounts array to be non-NULL
+		return calloc(1, sizeof(struct sc_mount));
+	}
+
+	int num_homedirs = 0;
+	for (char **path = inv->homedirs; *path != NULL; path++) {
+		num_homedirs++;
+	}
+
+	// We add one element for the end-of-array indicator.
+	struct sc_mount *mounts = calloc(num_homedirs + 1, sizeof(struct sc_mount));
+	if (mounts == NULL) {
+		die("cannot allocate mount data for homedirs");
+	}
+
+	struct sc_mount *current_mount = mounts;
+	for (char **path = inv->homedirs; *path != NULL; path++) {
+		debug("Adding homedir: %s", *path);
+		current_mount->path = *path;
+		current_mount++;
+	}
+	return mounts;
+}
+
+static void sc_free_dynamic_mounts(struct sc_mount *mounts)
```

Comment:
(super nitpick) I wonder if we should express in the function name that it does not do a "deep" free, maybe `sc_free_dynamic_mounts_shallow()` or `sc_free_dynamic_mounts_struct_only()` ? [and yay, rust to the rescue here]

Reply:
We now strdup and free them (per your other suggestion), so keeping name as-is!